packages/core/src/sql-on-fhir/eval.ts

```diff
@@ -0,0 +1,243 @@
+ * Processes a selection structure and node to emit all rows.
+ *
+ * This step emits all rows for a given Selection Structure and Node. We first generate sets of
+ * "partial rows" (i.e., sets of incomplete column bindings from the various clauses of V) and combine them to emit complete rows.
+ *
+ * This function is deliberately structured to match the pseudocode in the SQL-on-FHIR specification.
+ * See: https://build.fhir.org/ig/FHIR/sql-on-fhir-v2/implementer_guidance.html#processs-n-recursive-step
+ *
+ * @param s - The selection structure.
+ * @param n - The node (element) from a FHIR resource.
+ * @returns An array of output rows.
+ */
+function process(s: SelectionStructure, n: TypedValue): OutputRow[] {
+  const result: OutputRow[] = [];
+
+  // 1. Define a list of Nodes foci as
+  let foci: TypedValue[];
+  if (s.forEach) {
+    // If S.forEach is defined: fhirpath(S.forEach, N)
+    foci = evalFhirPathTyped(s.forEach, [n]);
+  } else if (s.forEachOrNull) {
+    // Else if S.forEachOrNull is defined: fhirpath(S.forEachOrNull, N)
+    foci = evalFhirPathTyped(s.forEachOrNull, [n]);
+  } else {
+    // Otherwise: [N] (a list with just the input node)
+    foci = [n];
+  }
+
+  // 2. For each element f of foci
+  for (const f of foci) {
+    // Initialize an empty list parts (each element of parts will be a list of partial rows)
+    const parts: OutputRow[][] = [];
+
+    // Process Columns:
+    for (const col of s.column ?? []) {
+      // For each Column col of S.column, define val as fhirpath(col.path, f)
+      const val = evalFhirPathTyped(col.path, [f]);
+
+      // Define b as a row whose column named col.name takes the value
+      let b: OutputRow;
+
+      if (val.length === 0) {
+        // If val was the empty set: null
+        b = { [col.name]: null };
+      } else if (val.length === 1) {
+        // Else if val has a single element e: e
+        b = { [col.name]: val[0].value };
+      } else if (col.collection) {
+        // Else if col.collection is true: val
+        b = { [col.name]: val.map((v) => v.value) };
+      } else {
+        // Else: throw "Multiple values found but not expected for column"
+        throw new Error('Multiple values found but not expected for column');
+      }
+
+      // Append [b] to parts
+      // (Note: append a list so the final element of parts is now a list containing the single row b).)
+      parts.push([b]);
+    }
+
+    // Process Selects:
+    // For each selection structure sel of S.select
+    for (const sel of s.select ?? []) {
+      // Define rows as the collection of all rows emitted by Process(sel, f)
+      const rows = process(sel, f);
+
+      // Append rows to parts
+      // (Note: do not append the elements but the whole list, so the final element of parts is now the list rows)
+      parts.push(rows);
+    }
+
+    // Process UnionAlls:
+    // Initialize urows as an empty list of rows
+    // For each selection structure u of S.unionAll
+    if (s.unionAll) {
```

Comment:
Should this be wrapped in the `if (s.unionAll)` check? Reading the pseudocode, it seems like you're still supposed to append an empty list if `S.unionAll` is missing. It may not matter given the cartesianProduct step below, but wanted to call it out.

Reply:
Great question, and astute observation.

My first stab did exactly that; however, it doesn't work.

The reason is that it ends with a call to `parts.push(urows);`  If `urows` is empty, you then have an empty array in `parts`. When calculating the cartesian product, an empty array is like multiplying by zero, it wipes out everything.

It could be handled by doing `if (urows.length > 0) { parts.push(urows); }`, or updating the cartesian product logic, but this felt cleaner 🤷